Add tests and document error handling for `CUDAPluggableAllocator`

emcastillo opened the pull request:
cc @shwina @leofang @albanD

We have seen several issues when dealing with errors in this setup as they may not be propagated accordingly,
Now, this function will return an error code and if it's `!=0` we will let torch fail on its side.

This is a breaking change as it changes the signature of the alloc/free external functions.

There are a bunch of issues when crossing the C/Python boundaries dealing with exceptions so we feel this PR is the "safest" way to deal with such situations.

Commit:
Handle errors for `CUDAPluggableAllocator`

leofang commented:
Glad to see you're converging to cuQuantum's design 😄 
https://docs.nvidia.com/cuda/cuquantum/cutensornet/api/types.html#_CPPv429cutensornetDeviceMemHandler_t


Minor question: Why is it a breaking change? Has the previous design been included in any public release?

emcastillo commented:
> Minor question: Why is it a breaking change? Has the previous design been included in any public release?
Not yet but I am not sure if the branch is already frozen

Commit:
lint

albanD commented on docs/source/notes/cuda.rst:
```diff
@@ -468,16 +468,17 @@ traces all the memory operations.
+   int my_free(void* ptr, ssize_t size, int device, cudaStream_t stream) {
       std::cout<<"free "<<ptr<< " "<<stream<<std::endl;
       cudaFree(ptr);
+      return err;
```
You need to get the cudaFree return above.

albanD commented on torch/csrc/cuda/CUDAPluggableAllocator.cpp:
```diff
@@ -91,7 +91,11 @@ void* CUDAPluggableAllocator::malloc(
+  void* r = nullptr;
+  int error = alloc_fn_(&r, size, device, stream);
+  TORCH_CHECK(
+      error != 0,
```
error == 0 if you want to fail when they're different.

albanD reviewed (commented):
Why is this prefered compared to the case where the user provided `my_malloc` would raise a reasonable directly? You can use `TORCH_CHECK()` there and I'm sure that, since you know where the error code comes from, you can provide an error message that is a lot better.

For 2.0, the branch cut is at the end of this week ;)

leofang commented:
> Why is this prefered compared to the case where the user provided my_malloc would raise a reasonable directly?

Maybe @emcastillo can explain his reasons, but to me this new change is desired (and was why we did it in cuQuantum) because we simply can't assume raising a C++ exception from a C interface is OK. Maybe it is for Torch, but from the C API perspective it's better to return an error code. The implementation providers can internally log in `my_malloc` if they have further info to share with users. 

On the RMM side, for example, it's been demo'd that people can plug in a "Python allocator" (which could raise Python exceptions and all that), and it's a huge help for debugging, say, memory jittering in a complex workflow. @shwina can share more detail.

albanD commented:
> On the RMM side, for example, it's been demo'd that people can plug in a "Python allocator" (which could raise Python exceptions and all that), and it's a huge help for debugging, say, memory jittering in a complex workflow. @shwina can share more detail.

I would be interested to hear more here!
How would the Python exception case work here? You say that it is ok to raise python exception from within `alloc_fn_`? Or that the user `alloc_fn_` that calls into Python should catch python errors and translate them into regular C-style errors?
In the second case, that sounds quite bad as the nice Python exception gets swallowed and converted to a basic C error which prints an int.

shwina commented:
Thanks for this @emcastillo!

I'm happy to help answer here from the RMM side. I do have some questions/comments of my own though:

1. I'm not sure the changes to the API are relevant/useful for RMM. [Our `allocate_fn` and `deallocate_fn` are C++ functions](https://github.com/rapidsai/rmm/blob/branch-23.04/python/rmm/_lib/torch_allocator.pyx), and we rely on Cython's C++->Python exception translation to propagate C++ errors up to the user. Furthermore, as Leo mentioned, our `allocate_fn` and `deallocate_fn` can run arbitrary Python code, and again, we rely on Cython "error return values" to ensure those are correctly propagated up to the user.

2. When hooking up the Torch allocator to RMM, we did run into a curious finding re: exception propagation. See discussion [here[(https://github.com/rapidsai/rmm/blob/branch-23.04/python/rmm/_lib/torch_allocator.pyx). I don't think we ever got to the bottom of it. Bringing it up here in case others here have seen something similar.

Commit:
fixes

emcastillo commented:
Yeah, this is an issue when we open a library using dlsym or ctypes, the exception can't correctly propagate through this boundary and we end up losing information. While this is an uncommon case, we feel like we should support this.

Commit:
lint

Commit:
lint

albanD commented:
> Yeah, this is an issue when we open a library using dlsym or ctypes, 

I am still confused: what is the right way to raise a nice error in your malloc/free?

emcastillo commented:
> I am still confused: what is the right way to raise a nice error in your malloc/free?

The problem happens when a .so file is dynamically loaded with dlsym/ctypes.
Errors being raised inside can't pass through the foreign function interface and we end up getting a regular sigabrt and losing all the relevant info. It is better to return an error code and let the main context fail.
(I am not sure if my explanation is 100% accurate, @leofang, @shwina )

albanD commented:
> The problem happens when a .so file is dynamically loaded with dlsym/ctypes.
Errors being raised inside can't pass through the foreign function interface and we end up getting a regular sigabrt and losing all the relevant info. It is better to return an error code and let the main context fail.

I think I'm missing some things (maybe due to some pure C code being involved?).
But all our lib here is based on having error handling in libtorch_python.so and c++ errors being raised in libtorch_cpu.so with both of these being loaded via ctypes by python code. So that can definitely be done.

emcastillo commented:
Even if the library is written in C++, if we raise an error inside the library. Since we are loading it through an [FFI](https://en.wikipedia.org/wiki/Foreign_function_interface) due to dlsym/ctypes being involved here. The exception does not make it to the main context and can't catch it or process it in any way. The user will only see an abort.

Probably I am also missing some important point in my explanation here

In this example by @kmaehashi 
```
int main() {
    void* lib = dlopen("./lib.so", RTLD_LAZY);
    int (*fptr)();
    //fptr = (int(*)())dlsym(lib, "myfunc1");
    try {
        fptr = (int(*)())dlsym(lib, "myfunc2");
    } catch(...) {
        printf("error\n");
    }
    int out = fptr();
    printf("%d\n", out);
}
```

`myfunc2` code is
```
extern "C" int myfunc2() {
    throw std::invalid_argument("error");
}
```

However, compiling the myfunc2 as a .so and running the main here doesn't catch the exception. It just 
dies

```
terminate called after throwing an instance of 'std::invalid_argument'
  what():  error
[1]    756574 abort (core dumped)  ./test
```

emcastillo commented:
Any comments on this?

albanD commented:
Hey!

The example above has the try/catch around loading the function pointer (which doesn't throw as expected) but is not protecting the call itself `    int out = fptr();`. Adding a try/catch around the call does catch the error as expected for me when compiling this with g++.

leofang commented:
The way I see it, the concern is mainly that the current allocator interface assumes `alloc_fn` and `free_fn` are implemented in C++ and can raise C++ exceptions (or have other undocumented means for error handling), but this is not necessary the case. For example, one can imaging we just dlopen libcudart and attach `cudaMallocAsync` and `cudaFreeAsync` as is, without any additional C++ shim layer to convert the CUDA error code to a C++ exception. The current interface does not accept that. Another use case could be using CuPy's Cython based allocator, which does not throw C++ exception either. I am sure there are more allocator usage in the wild that do not have a C++ shim layer.

malfet commented:
Few clarification on exception handling across C/C++ shared libraries:
 - `extern "C"` has no effect on the runtime behavior on the function/class/variable, but only on how it is linked. 
 - If shared C++ runtime is used, than dynamic library decomposition of app is irrelevant, assuming all components are adhere to the same ABI. That is exception raised from `a.cpp` can be caught in `b.cpp` regardless of whether they are linked together as single executable, or if executable and shared library. But executable compiled with CXX11 ABI might not catch the exception raised from library compiled with pre-CXX11 ABI and vice versa.
 - Pure C function are transparent to exception handling, which might cause some resource leaks.
 
I guess what proposed PR is missing, is clear documentation of how errors are reported: i.e. are `alloc_fn` behaves like `new` (that raises `std::bad_alloc` exception on failures) or like `malloc` (that returns null on failure unless alloc size is 0)

leofang commented:
Yes, in any case a clear documentation on expected exception handling is needed. The question is what the error behavior should we expect for `alloc_fn` and `free_fn`, which is currently undocumented and gives room for potential change.

emcastillo commented:
Rebased on top of master, will add tests and work on the docs too, thanks!!

emcastillo commented:
/test

Commit:
minor fixes

Commit:
Add tests

Commit:
add docs

Commit:
Do not run on windows or rocm

emcastillo commented:
Added tests and added some more info on docs section

Commit:
use exceptions

Commit:
Error with exceptions

emcastillo commented:
I reverted this PR to the previous error handling but added testing and modified docs.
Tests show that the exception and `TORCH_CHECK`works fine. However, I think that the real issues here were the interoperability with Cython cc @shwina @leofang

albanD commented on docs/source/notes/cuda.rst:
```diff
@@ -511,6 +511,61 @@ of the alloc/free functions that match the signatures specified above.
+    }
+    
+    // Needed for torch extension to work
+    PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {}
```
You should provide the `functions=` argument to load_inline to ensure they are properly exposed?

emcastillo replied:
adding that argument generates python bindings for the functions.
But we want the C++ function address to hook it in the c10 allocator, so exposing them won't give us this pointer :(

albanD replied:
It will expose them to the python side but the original will also be available on the .so ?

albanD commented on test/test_cuda.py:
```diff
@@ -5770,6 +5770,189 @@ def test_no_triton_on_import(self):
+    cpp_sources=custom_allocator_code,
+    build_directory='.',
+    with_cuda=True,
+)
```
I got the same error locally as in CI: no .so is generated because you don't ask for any functions.

emcastillo replied:
This is weird, I see all the .so functions generated in my local env ...
Could this be compiler versions? or the path where the compiled module is being stored?

emcastillo replied:
after building them in a tempdir this was solved :)

albanD commented on test/test_cuda.py:
```diff
@@ -5770,6 +5770,189 @@ def test_no_triton_on_import(self):
+                cwd=os.path.dirname(os.path.realpath(__file__))).strip().decode('ascii')
+        except subprocess.CalledProcessError as e:
+            expected_message = "bad custom free"
+            print(e.output.decode)
```
nit print

albanD commented on test/test_cuda.py:
```diff
@@ -5770,6 +5770,189 @@ def test_no_triton_on_import(self):
+            expected_message = "bad custom free"
+            print(e.output.decode)
+            self.assertTrue(expected_message in e.output.decode('ascii'))
+            error = True
+        self.assertTrue(error)
```
nit
```suggestion
        else:
            self.assertTrue(False, "Failed to fail")
```

albanD reviewed (commented):
The example and CI needs fixing. But change sounds ok to me otherwise.

Curious to hear more if that does not work well with Cython though?

Commit:
fixes

Commit:
lint & tempdir

albanD commented on docs/source/notes/cuda.rst:
```diff
@@ -511,6 +511,61 @@ of the alloc/free functions that match the signatures specified above.
+
+    custom_allocator_code = """
+    #include <sys/types.h>
+    #include <cuda_runtime_api.h>
```
I don't think this include is needed right? It is already added by default.

albanD reviewed (approved):
Sounds good!
Only small nits

emcastillo commented:
For the Cython interoperability I managed to find an interesting example:

Lets take the following CuPy PR where we expose the allocator :)

https://github.com/cupy/cupy/pull/7210

Basically I just modify this PR to throw an exception in the malloc function
```
cdef public void* cupy_malloc_ext(
        ssize_t size, int device_id, void* stream) except * with gil:
        raise RuntimeError('error on malloc')
```
If I now load the CuPy generated .so file with the `CUDAPluggableAllocator` interface in PyTorch as in the example code in the PR
```
cupy_allocator_so = 'cupy/cuda/memory.cpython-39-x86_64-linux-gnu.so'
new_alloc = torch.cuda.memory.CUDAPluggableAllocator(cupy_allocator_so, 'cupy_malloc_ext', 'cupy_free_ext')

torch.cuda.memory.change_current_allocator(new_alloc)
```


However, with PyTorch in release mode (install with pip install torch==2.0.0)

```
new allocator changed
Traceback (most recent call last):
  File "/home/ecastill/em-cupy/tor.py", line 13, in <module>
    print(torch.zeros(10, device='cuda'))
RuntimeError: CUDA error: an illegal memory access was encountered
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

The exception is ignored and a value of 0 is returned :) I am not sure if this is a Cython, dlopen or Pytorch issue.


However, with pytorch compiled in development mode, the exception is correctly intercepted

```
new allocator changed
Malloc called and returned 0
Traceback (most recent call last):
  File "cupy/cuda/memory.pyx", line 1982, in cupy.cuda.memory.cupy_malloc_ext
RuntimeError: error on malloc
```

IIRC this is what motivated us to use error codes!

kmaehashi commented:
> Curious to hear more if that does not work well with Cython though?

This issue isn't solely related to Cython or CuPy. Consider a situation where an allocator is written in Rust or any other language that's incapable of emitting C++ exceptions. The existing `CUDAPluggableAllocator` API doesn't permit such third-party allocators to convey errors to PyTorch without having to write additional glue code to convert an error into a C++ exception. It might be a better idea to remove the expectation for C++ exceptions at the C-linkage API boundary to improve interoperability.

albanD commented:
Thanks for the full picture on this!
Would a final proposal then be to have the malloc signature updated to return an optional error code.
We recommend users always return 0 code and raise c++ errors when an error happens (if they binding allows it).
When using C-API, they can also return a non-0 error code that will raise a generic error with the user with the error code.

Would that give us the best of both worlds?
